flink-connector-aws/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/config/RecoverableErrorsConfig.java

```diff
@@ -0,0 +1,99 @@
+import java.util.List;
+import java.util.Optional;
+import java.util.Properties;
+
+/**
+ * Hosts the recoverable exception configuration. Recoverable exceptions are retried indefinitely.
+ */
+public class RecoverableErrorsConfig {
+    public static final String INVALID_CONFIG_MESSAGE =
+            "Invalid config for recoverable consumer exceptions. "
+                    + "Valid config example: "
+                    + "`flink.shard.consumer.error.recoverable[0].exception=net.java.UnknownHostException`. "
+                    + "Your config array must use zero-based indexing as shown in the example.";
+
+    /**
+     * Parses the array of recoverable error configs.
+     *
+     * @param config connector configuration
+     * @return an Optional of RecoverableErrorsConfig
+     */
+    public static Optional<RecoverableErrorsConfig> createConfigFromPropertiesOrThrow(
+            final Properties config) {
+        List<ExceptionConfig> exConfs = new ArrayList<>();
+        int idx = 0;
+        String exceptionConfigKey =
+                String.format(
+                        "%s[%d].exception",
+                        ConsumerConfigConstants.RECOVERABLE_EXCEPTIONS_PREFIX, idx);
+        while (config.containsKey(exceptionConfigKey)) {
+            String exPath = config.getProperty(exceptionConfigKey);
+            try {
+                Class<?> aClass = Class.forName(exPath);
+                if (!Throwable.class.isAssignableFrom(aClass)) {
+                    throw new ClassCastException();
+                }
+                exConfs.add(new ExceptionConfig(aClass));
+            } catch (ClassCastException e) {
+                throw new IllegalArgumentException(
+                        "Provided recoverable exception class is not a Throwable: " + exPath);
+            } catch (ClassNotFoundException e) {
+                throw new IllegalArgumentException(
+                        "Provided recoverable exception class could not be found: " + exPath);
+            }
+            exceptionConfigKey =
+                    String.format(
+                            "%s[%d].exception",
+                            ConsumerConfigConstants.RECOVERABLE_EXCEPTIONS_PREFIX, ++idx);
+        }
+        if (idx > 0) {
+            // We processed configs successfully
+            return Optional.of(new RecoverableErrorsConfig(exConfs));
+        }
+
+        // Check if user provided wrong config suffix, so they fail faster
+        for (Object key : config.keySet()) {
+            if (((String) key).startsWith(ConsumerConfigConstants.RECOVERABLE_EXCEPTIONS_PREFIX)) {
+                throw new IllegalArgumentException(RecoverableErrorsConfig.INVALID_CONFIG_MESSAGE);
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    private final List<ExceptionConfig> exceptionConfigs;
+
+    public RecoverableErrorsConfig(List<ExceptionConfig> exceptionConfigs) {
+        this.exceptionConfigs = exceptionConfigs;
+    }
```

Comment:
Please move constructors and fields above methods, this is best practise for java.